### PR TITLE
Add SnippetDataProvider for SmartContent with Audience Targeting

### PR DESCRIPTION
**Pull request by danrot**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes parts of #3205 
| Related issues/PRs | #issuenum
| License | MIT
| Documentation PR | ---

#### What's in this PR?

This PR adds an excerpt tab to the snippets and a snippet smart content provider. The smart content provider will also allow to take audience targeting into account.

### CODE REVIEW

File: src/Sulu/Bundle/SnippetBundle/Content/SnippetQueryBuilder.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ */
+class SnippetQueryBuilder extends QueryBuilder
+{
+    protected static $mixinTypes = ['sulu:snippet'];
+
+    public function buildWhere($webspaceKey, $locale)
+    {
+        $sqlWhere = parent::buildWhere($webspaceKey, $locale);
+        if ($types = $this->getConfig('types')) {
+            $sqlWhere .= ' AND (' . implode(' OR ', array_map(function($type) {
```
Comment (danrot) [inline src/Sulu/Bundle/SnippetBundle/Content/SnippetQueryBuilder.php]: try to move that query logic into the already existing path

**Comment by chirimoya** [comment]
Property mapping is not working as expected

```xml
        <property name="snippets" type="smart_content">
            <meta>
                <title lang="de">Snippets</title>
                <title lang="en">Snippets</title>
            </meta>

            <params>
                <param name="provider" value="snippet"/>
                <param name="types" value="msg"/>

                <param name="properties" type="collection">
                    <param name="description" value="description"/>
                </param>
            </params>
        </property>
```

**Comment by danrot** [comment]
The `excerpt` should only contain tags, categories and the audience targeting groups for the snippets.

**Comment by danrot** [comment]
Everything should be working now, except for the access of the `resource` in the smart content. This is a PHP7 issue, which was fixed in a new major release of the ProxyManager (see https://github.com/Ocramius/ProxyManager/issues/319 for more details). We should probably upgrade, that's the only fix for this error. Could also be done in a separate PR, or even in the master branch (since the `resource` variable is not working for any data provider).

**Comment by danrot** [comment]
@chirimoya I fixed the double delete dialog already in this PR, it's the last commit.

**Pull request merged**